src/cc/BPF.cc

```diff
@@ -0,0 +1,418 @@
+/*
+ * Copyright (c) 2016 Facebook, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include <cstdio>
+#include <exception>
+#include <iostream>
+#include <linux/bpf.h>
+#include <memory>
+#include <sstream>
+#include <unistd.h>
+#include <vector>
+
+#include "bcc_syms.h"
+#include "bpf_module.h"
+#include "common.h"
+#include "exception.h"
+#include "libbpf.h"
+#include "perf_reader.h"
+
+#include "BPF.h"
+
+namespace ebpf {
+
+std::string uint_to_hex(uint64_t value) {
+  std::stringstream ss;
+  ss << std::hex << value;
+  return ss.str();
+}
+
+std::string sanitize_str(
+  std::string str, bool (*validator)(char), char replacement = '_'
+) {
+  for (size_t i = 0; i < str.length(); i++)
+    if (!validator(str[i]))
+      str[i] = replacement;
+  return str;
+}
+
+BPF::BPF(
+  const std::string& bpf_program,
+  unsigned int debug_flag,
+  std::vector<std::string> cflags
+) :
+  bpf_module_(new BPFModule(debug_flag))
+{
+  const char* flags[cflags.size()];
+  for (size_t i = 0; i < cflags.size(); i++)
+    flags[i] = cflags[i].c_str();
+  if (bpf_module_->load_string(bpf_program, flags, cflags.size()) != 0)
+    throw std::runtime_error("Unable to initialize BPF program");
+};
+
+BPF::~BPF() {
+  for (auto it : kprobes_) {
+    auto res = detach_kprobe_event(it.first, it.second);
+    if (std::get<0>(res) != 0)
+      std::cerr << "Failed to detach kprobe event " << it.first
```

Comment:
The idea here is that if we are failing in destructor, then there would be nothing else we could do, so I'd rather inform the user to let them know. But I see your point, will move this out so people could do clean up whenever they want.